Create rule S6611 (#1799)
SonarSource/rspec: rules/S6611/kotlin/rule.adoc

The type system in Kotlin is designed to eliminate the risk of null pointer references by distinguishing between references that can be null and references that cannot be null. 

== Why is this an issue?

> It's not always clear whether a variable is guaranteed to be non-null or not so it's easy to misuse this operator. If a null value is passed to a function that expects a non-null value, it can lead to a crash or unexpected behavior.

Reviewer: Actually, in Kotlin it should be NullPointerException if we pass `null` where `null` is not expected, but if we have Java interop it could be also an unexpected behavior. Maybe it's better to mention this explicitly.

Also I've just realized that the rule is actually about accessing `Map` values and we don't talk about `Map` in the description. So we need something like: "When accessing map elements by key, we can't be sure that value exists so we should handle the `null` value properly"

== How to fix it

> Avoid using the non-null assertion operator.

Reviewer: `Map` part is missing here. 
```suggestion
Avoid using the non-null assertion operator when accessing `Map` elements.
```

Reviewer: And let's add a few ideas on what can be used instead: 
- just return an instance of a nullable type
- use `getValue`, this will throw `NoSuchElementException` when an element is not present
- use `getOrElse`/`getOrDefault` and specify the behavior in case of `null` explicitly
- use am elvis operator `?:` to specify the behavior in case of `null`

> Using the non-null assertion operator in Kotlin, denoted by the exclamation mark (!) can lead to unexpected runtime errors if the value turns out to be null. It's easy to misuse the operator, and it's not always clear whether a variable is guaranteed to be non-null or not. If a null value is passed to a function that expects a non-null value, it can lead to a crash or unexpected behavior.

Reviewer: it's actually a double axclamation mark `!!`
and `unexpected runtime errors` - it's actually expected NullPointerException

Reviewer 2: I took out that sentence, I think it works better without it

=== Code examples

==== Noncompliant code example

This example will cause an error because the variable is actually nullable. 

[source,kotlin]
----
val l = mapOf(1 to "one", 2 to "two", 3 to "five")
l.get(123)!! // Noncompliant
l[123]!! // Noncompliant
----

==== Compliant solution

By removing the non-null assertion operator the compiler is able to return nullable when necessary.

[source,kotlin]
----
val l = mapOf(1 to "one", 2 to "two", 3 to "five")
l.get(123) // Compliant, returns nullable
l[123] // Compliant, returns nullable

> l.getValue(123) // Compliant, throws NoSuchElemnException

Reviewer: `throws NoSuchElementException`

When accessing `Map` elements by key, you should handle the `null` value properly in case the value does not exist.

== Why is this an issue?

It's not always clear whether a variable is guaranteed to be non-null or not so it's easy to misuse this operator. In Kotlin if you pass null when a null value is not expected, you will get a NullPointerException. However, if you have a Java interop then passing an expected null value can result in unexpected behavior. 

== How to fix it

You should avoid using the non-null assertion operator when accessing `Map` elements. 

Instead, you can:

* return an instance of a nullable type.
* use getValue. This will throw NoSuchElementException when an element is not present.
* use either `getOrElse` or `getOrDefault` functions and explicitly specify the behavior in case of a null value.

> * use am elvis operator ?: to specify the behavior in case of a null value.

Reviewer: ```suggestion
* use an elvis operator `?:` to specify the behavior in case of a null value.
```

=== Code examples

==== Noncompliant code example

> This example will cause an error because the variable is actually nullable. 

Reviewer: ```suggestion
This example will throw a NullPointerException at the `!!` operator because `map` doesn't have a key `123` and the result of `get(123)` will be null. 
```

[source,kotlin]
----
val l = mapOf(1 to "one", 2 to "two", 3 to "five")
l.get(123)!! // Noncompliant
l[123]!! // Noncompliant
----

==== Compliant solution

> By removing the non-null assertion operator the compiler is able to return nullable when necessary.

Reviewer: ```suggestion
By removing the non-null assertion operator the result of the `get(123)` call will return an instance of a nullable type, and the user can handle the potential `null` value properly
```

> == Why is this an issue?

Reviewer 3: I think we should add a code snippet to this section

> == Why is this an issue?

Reviewer 3: From this section I don't understand what we try to detect

> When accessing `Map` elements by key, you should handle the `null` value properly in case the value does not exist. It's not always clear whether a variable is guaranteed to be non-null or not so it's easy to misuse the non-null operator. 

Reviewer 3: Which kind of misuse are we talking about?
Which kind of variable are we taking about in `It's not always clear whether a variable is guaranteed to be non-null or not`?

> In Kotlin if you pass a null when a null value is not expected, you will get a NullPointerException. However, if you have a Java interop then passing an expected null value can result in unexpected behavior. 

Reviewer 3: Which kind of unexpected behaviour?

== How to fix it

You should avoid using the non-null assertion operator when accessing `Map` elements. 

Instead, you can:

* return an instance of a nullable type.

> * use getValue. This will throw NoSuchElementException when an element is not present.

Reviewer 3: ```suggestion
* use `getValue`. This will throw `NoSuchElementException` when an element is not present.
```
There are more places where backticks are missing

* use either `getOrElse` or `getOrDefault` functions and explicitly specify the behavior in case of a null value.
* use an elvis operator `?:` to specify the behavior in case of a null value.

=== Code examples

==== Noncompliant code example

This example will throw a NullPointerException at the `!!` operator because `map` doesn't have a key `123` and the result of `get(123)` will be null. 

[source,kotlin]
----
val l = mapOf(1 to "one", 2 to "two", 3 to "five")

> l.get(123)!! // Noncompliant

Reviewer 3: How about adding for example print statement to make this code more realistic?

Reviewer: I am not sure it will help understand the rule

l[123]!! // Noncompliant
----

==== Compliant solution

> By removing the non-null assertion operator the result of the `get(123)` call will return an instance of a nullable type, and the user can handle the potential `null` value properly

Reviewer 3: ```suggestion
By removing the non-null assertion operator the result of the `get(123)` call will return an instance of a nullable type, and the user can handle the potential `null` value properly.
```
